keps/sig-autoscaling/4943-vpa-heterogeneous-workloads/README.md

```diff
@@ -0,0 +1,946 @@
+broadly suggests that workloads be given more capacity on larger nodes. That
+heuristic may be good enough for some use cases, and this proposal implicitly
+allows for the possibility of extending the "scope" field to support such a
+feature in the future. However, this proposal may also address the same
+underlying need more precisely. Conversely, scaling based on node metadata
+alone would not meet the needs of some of the motivating use cases for this
+proposal, because the underlying assumption of scaling based on node metadata
+is that the workload scales directly proportionally to the node.  We know this
+is not the case in all circumstances.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+One significant risk is scaling this feature. While VPA generally scales in a
+sublinear way because only a single recommendation is calculated, a VPA with a
+node scope will scale linerarly, requiring separate metrics queries for each
+node and separate recommendation calculations. Care will be taken to ensure
+this is handled efficiently.
+
+As of Kubernetes V1.32, [up to 5,000 nodes are
+supported](https://kubernetes.io/docs/setup/best-practices/cluster-large/) per
+cluster. There is no official recommendation on the maximum number of
+VerticalPodAutoscalers that are supported. However, this proposal does increase
+the load for VerticalPodAutoscalers that use the new node scope option.
+
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.  -->
+
+First, the implementation of this feature adds the `Scope` field to the
+`VerticalPodAutoscalerSpec` struct:
+
+```
+type VerticalPodAutoscalerSpec struct {
+	TargetRef *autoscaling.CrossVersionObjectReference `json:"targetRef" protobuf:"bytes,1,name=targetRef"`
+
+	UpdatePolicy *PodUpdatePolicy `json:"updatePolicy,omitempty" protobuf:"bytes,2,opt,name=updatePolicy"`
+
+	ResourcePolicy *PodResourcePolicy `json:"resourcePolicy,omitempty" protobuf:"bytes,3,opt,name=resourcePolicy"`
+
++   // Scope indicates at what level the recommendations should be calculated and applied.
++   // By default, the scope is all pods under the controller specified in `TargetRef`. Scope
++   // may also be set to "node" when the TargetRef is a DaemonSet, to apply recommendations
++   // independently to the pod scheduled to each node under the DaemonSet.
++   Scope string `json:"scope,omitzero" protobuf:"bytes,4,opt,name=scope"`
```

Comment:
I wonder if we should have an enum here instead...

Did you pull this scope field as inspiration from somewhere? If so we definitely should be consistent, otherwise a more rigid API via an enum might be better here...